File: src/main/java/org/trp/shincolle/entity/base/ShipRecoveryDecisionResolver.java
package org.trp.shincolle.entity.base;

final class ShipRecoveryDecisionResolver {
    private ShipRecoveryDecisionResolver() {
    }

    static boolean shouldAttemptTeleport(State state) {
        return state.force() || state.distanceSqr() > state.teleportDistanceSqr();
    }

    static boolean shouldClearAfterStuck(int stuckTicks, int stuckTickLimit) {
        return stuckTicks > stuckTickLimit;
    }

    static boolean shouldClearAfterMoveFailures(int failCount, int failLimit) {
        return failCount > failLimit;
    }

    record State(
            boolean force,
            double distanceSqr,
            double teleportDistanceSqr
    ) {
    }
}

File: src/main/java/org/trp/shincolle/entity/base/ShipRecoveryDecisionResolver.kt
package org.trp.shincolle.entity.base

internal object ShipRecoveryDecisionResolver {
    fun shouldAttemptTeleport(state: State): Boolean {
        return state.force || state.distanceSqr > state.teleportDistanceSqr
    }

    fun shouldClearAfterStuck(stuckTicks: Int, stuckTickLimit: Int): Boolean {
        return stuckTicks > stuckTickLimit
    }

    fun shouldClearAfterMoveFailures(failCount: Int, failLimit: Int): Boolean {
        return failCount > failLimit
    }

    @JvmRecord
    internal data class State(
        val force: Boolean,
        val distanceSqr: Double,
        val teleportDistanceSqr: Double
    )
}
